File: src/main/java/com/syfe/personalfinance/controller/AuthController.java
package com.syfe.personalfinance.controller;

import com.syfe.personalfinance.dto.AuthDto;
import com.syfe.personalfinance.dto.AuthDto.UserResponse;
import com.syfe.personalfinance.service.UserService;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpSession;
import jakarta.validation.Valid;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.web.bind.annotation.*;

@RestController
@RequestMapping("/api/auth")
public class AuthController {

    private final UserService userService;
    private final AuthenticationManager authenticationManager;

    public AuthController(UserService userService, AuthenticationManager authenticationManager) {
        this.userService = userService;
        this.authenticationManager = authenticationManager;
    }

    @PostMapping("/register")
    public ResponseEntity<AuthDto.RegisterResponse> register(@Valid @RequestBody AuthDto.RegisterRequest request) {
        UserResponse responseData = userService.register(request);
        AuthDto.RegisterResponse response = AuthDto.RegisterResponse.builder()
                .message("User registered successfully")
                .userId(responseData.getId())
                .build();
        return new ResponseEntity<>(response, HttpStatus.CREATED);
    }

    @PostMapping("/login")
    public ResponseEntity<AuthDto.SimpleMessageResponse> login(@Valid @RequestBody AuthDto.LoginRequest request, HttpServletRequest servletRequest) {
        UsernamePasswordAuthenticationToken authToken = new UsernamePasswordAuthenticationToken(
                request.getUsername(), request.getPassword());

        Authentication authentication = authenticationManager.authenticate(authToken);
        SecurityContextHolder.getContext().setAuthentication(authentication);

        HttpSession session = servletRequest.getSession(true);
        session.setAttribute("SPRING_SECURITY_CONTEXT", SecurityContextHolder.getContext());

        AuthDto.SimpleMessageResponse response = AuthDto.SimpleMessageResponse.builder()
                .message("Login successful")
                .build();

        return ResponseEntity.ok(response);
    }

    @PostMapping("/logout")
    public ResponseEntity<AuthDto.SimpleMessageResponse> logout(HttpServletRequest servletRequest) {
        HttpSession session = servletRequest.getSession(false);
        if (session != null) {
            session.invalidate();
        }
        SecurityContextHolder.clearContext();

        AuthDto.SimpleMessageResponse response = AuthDto.SimpleMessageResponse.builder()
                .message("Logout successful")
                .build();

        return ResponseEntity.ok(response);
    }
}

File: src/main/java/com/syfe/personalfinance/controller/AuthController.kt
package com.syfe.personalfinance.controller

import com.syfe.personalfinance.dto.AuthDto
import com.syfe.personalfinance.service.UserService
import jakarta.servlet.http.HttpServletRequest
import jakarta.validation.Valid
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.security.authentication.AuthenticationManager
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController

/**
 * REST controller managing user authentication lifecycle: registration, login session establishment, and logout.
 */
@RestController
@RequestMapping("/api/auth")
class AuthController(
    private val userService: UserService,
    private val authenticationManager: AuthenticationManager
) {

    /**
     * Registers a new user account with validated credentials.
     */
    @PostMapping("/register")
    fun register(@Valid @RequestBody request: AuthDto.RegisterRequest): ResponseEntity<AuthDto.RegisterResponse> {
        val responseData = userService.register(request)
        val response = AuthDto.RegisterResponse(
            message = "User registered successfully",
            userId = responseData.id
        )
        return ResponseEntity(response, HttpStatus.CREATED)
    }

    /**
     * Authenticates user credentials and establishes an HTTP session (JSESSIONID cookie).
     */
    @PostMapping("/login")
    fun login(
        @Valid @RequestBody request: AuthDto.LoginRequest,
        servletRequest: HttpServletRequest
    ): ResponseEntity<AuthDto.SimpleMessageResponse> {
        val authToken = UsernamePasswordAuthenticationToken(request.username, request.password)

        val authentication = authenticationManager.authenticate(authToken)
        SecurityContextHolder.getContext().authentication = authentication

        val session = servletRequest.getSession(true)
        session.setAttribute("SPRING_SECURITY_CONTEXT", SecurityContextHolder.getContext())

        val response = AuthDto.SimpleMessageResponse(
            message = "Login successful"
        )

        return ResponseEntity.ok(response)
    }

    /**
     * Terminates the current HTTP session and clears security context.
     */
    @PostMapping("/logout")
    fun logout(servletRequest: HttpServletRequest): ResponseEntity<AuthDto.SimpleMessageResponse> {
        val session = servletRequest.getSession(false)
        session?.invalidate()
        SecurityContextHolder.clearContext()

        val response = AuthDto.SimpleMessageResponse(
            message = "Logout successful"
        )

        return ResponseEntity.ok(response)
    }
}
